demos/object_detection_demo_ssd_async/main.cpp

```diff
@@ -354,58 +401,197 @@ int main(int argc, char *argv[]) {
-            }
+                InferRequest::Ptr request;
+                {
+                    std::lock_guard<std::mutex> lock(mutex);
 
-            // Final point:
-            // in the truly Async mode we swap the NEXT and CURRENT requests for the next iteration
-            curr_frame = next_frame;
-            next_frame = cv::Mat();
-            if (isAsyncMode) {
-                async_infer_request_curr.swap(async_infer_request_next);
-            }
+                    request = std::move(emptyRequests.front());
+                    emptyRequests.pop_front();
+                }
+                frameToBlob(frame, request, imageInputName);
+                
+                ExecutionMode frameMode = currentMode;
+                request->SetCompletionCallback([&mutex,
+                                                &completedRequestResults,
+                                                nextFrameId,
+                                                frame,
+                                                request,
+                                                &outputName,
+                                                startTime,
+                                                frameMode,
+                                                &currentMode,
+                                                &emptyRequests,
+                                                &callbackException,
+                                                &condVar] {
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    std::make_shared<TBlob<float>>(*as<TBlob<float>>(request->GetBlob(outputName))),
+                                    startTime,
+                                    frameMode == currentMode
+                                }));
+                            
+                            emptyRequests.push_back(std::move(request));
+                        }
+                        catch (...) {
+                            if (!callbackException) {
+                                callbackException = std::current_exception();
+                            }
+                        }
+                    }
+                    condVar.notify_one();
+                });
 
-            const int key = cv::waitKey(1);
-            if (27 == key)  // Esc
-                break;
-            if (9 == key) {  // Tab
-                isAsyncMode ^= true;
-                isModeChanged = true;
+                request->StartAsync();
+                nextFrameId++;
             } else {
-                presenter.handleKey(key);
+                while (true) {
+                    std::unique_lock<std::mutex> lock(mutex);
```

Comment:
Why did you move the lock inside the loop? There's no reason to repeatedly relock the mutex, the original loop structure was fine.